### PR TITLE
Adds a new salvager weapon: The Makeshift Laser Rifle

### PR DESCRIPTION
**Pull request by unfunnyperson**
[FEATURE] [GAME OBJECTS] [INPUT] [WIKI]

## About the PR 

Adds a new salvager weapon!
image

The weapon is a 2 handed laser weapon.
It fits on backs as well.
It uses large energy cells and uses 1250 energy per shot to create a 20 damage laser.

different cells used
image
image

The weapon can be used in-hand to adjust the firerate. This is between 0.2 and 1 seconds
image

If the gun is fired too much within a short period of time or the gun is emped the gun will burst into flame and the gun's cell will be drained!
image
After cooling down the gun needs to be repaired
image
Steps:
1. Wirecutters to remove burnt wires
2. Add 10 wires to replace burnt wires
3. remove and replace power cell/light tube

The meltdown combined with the adjustable firerate allows a player to choose a slower firerate for a more reliable gun or a faster firerate for higher dps.

Using a rigged light tube will cause the gun to explode on use.

Using a wrench on the rifle will remove the battery
Using wirecutters will remove the light tube

A new battery/light tube can be added to the gun by clicking on it with the respective item

## Why's this needed? 

More variety in salvager weapons provides a more interesting gameplay experience. This weapon is designed around being less powerful than the survival rifle but has more ammo sources than the rifle as compensation. 

## Changelog

```changelog
(u)UnfunnyPerson
(+)Added a new salvager weapon, the makeshift laser rifle.
```

### CODE REVIEW

**Comment by unfunnyperson** [comment]
Still need to add help messages and a sprite for the potato cell

**Comment by pgmzeta** [comment]
it would be cool if it used the matsci parts like the captain's gun does

**Comment by CheffieGithub** [comment]
On testing I think you could up the damage, on the fastest fire rate it was able to fire 7 times for 84 damage. On the slowest 12 times for 144 damage. Maybe change the temperature scaling a bit as well?

Considering the RP-3 micro phaser is incredibly easy to get and can do more damage in about the same span of time with 10 damage and 20 shots with dual wield capabilities, pocket storage and easier cell swapping.

File: code/obj/item/gun/energy.dm [hunk 1]
```diff
@@ -1858,3 +1858,276 @@ TYPEINFO(/obj/item/gun/energy/wasp)
...
+			our_light = W
+			W.set_loc(src)
+			update_icon()
+			return
+		..()
+
+	examine()
+		. = ..()
+		if (!our_cell)
+			. += "[src] is lacking a power source!"
```
Comment (CheffieGithub) [inline code/obj/item/gun/energy.dm:1984]: I would use `get_desc()` here instead and add cases for missing / broken tube.

Comment (CheffieGithub) [reply code/obj/item/gun/energy.dm:1984]: there should also be a space before `[src]` here

**Comment by unfunnyperson** [comment]
Changed damage from 12 >18
Added a new construction step to make construction a little more difficult
The gun overheating is now a lot worse and need repairs

**Comment by imawxome** [comment]
i really like this, its great to have some usefull makeshift guns

**Comment by pali6** [comment]
I personally am not convinced we want something like a makeshift laser gun in the game.

File: code/obj/item/gun/energy.dm [hunk 2]
```diff
@@ -1839,3 +1839,428 @@ TYPEINFO(/obj/item/gun/energy/wasp)
...
+	///What light source we use for the rifle
+	var/obj/item/light/tube/our_light
+	///What battery this gun uses
+	var/obj/item/cell/our_cell
+	///How much heat this weapon has after firing, the weapon breaks if this gets too high
+	var/heat = 0
+	///What step of repair are we on if we have broken? 0 = functional
+	var/heat_repair = 0
+	///What projectile we switch to when update_proj() is called
+	var/lens_proj = /datum/projectile/laser/makeshift
```
Comment (Azrun) [inline code/obj/item/gun/energy.dm:1878]: Why is this necessary given the existing members on this type?

**Comment by unfunnyperson** [comment]
reopening this because I want to make this a weapon for salvagers instead of a makeshift crew weapon

File: code/obj/item/gun/energy.dm (hunk 2 above)
Comment (unfunnyperson) [reply code/obj/item/gun/energy.dm]: fixed

**Comment by unfunnyperson** [comment]
don't have enough time to fix this right now, ill try to get it done on Friday or Thursday

File: code/obj/item/gun/energy.dm [hunk 3]
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
...
+	var/heat_repair = 0
+
+	proc/attach_cell(var/obj/item/cell/C, mob/user)
+		if (user)
+			user.u_equip(C)
+		our_cell = C
+		our_cell.set_loc(src)
+		our_cell.AddComponent(/datum/component/power_cell, our_cell.maxcharge, our_cell.charge, our_cell.genrate, 0, FALSE)
+		SEND_SIGNAL(src, COMSIG_CELL_SWAP, our_cell)
+		update_icon()
```
Comment (Azrun) [inline code/obj/item/gun/energy.dm:1884]:
```suggestion
		UpdateIcon()
```

File: code/obj/item/gun/energy.dm [hunk 4]
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
...
+	New()
+		processing_items |= src
+		set_current_projectile(new/datum/projectile/laser/makeshift)
+		projectiles = list(current_projectile)
+		..()
+
+	Exited(Obj, newloc)
+		var/obj/item/cell/C = Obj
+		if (istype(C) && !QDELETED(C))
+			C.update_icon()
```
Comment (Azrun) [inline code/obj/item/gun/energy.dm:1935]:
```suggestion
			C.UpdateIcon()
```

File: code/obj/item/gun/energy.dm [hunk 5]
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
...
+				user.u_equip(W)
+				boutput(user,SPAN_NOTICE("You attach [W] to [src]."))
+				attach_cell(W, user)
+				return
+			else if (issnippingtool(W) && our_light)
+				boutput(user,SPAN_NOTICE("You remove [our_light] from the barrel."))
+				playsound(src, 'sound/items/Wirecutter.ogg', 50, TRUE)
+				user.put_in_hand_or_drop(our_light)
+				our_light = null
+				update_icon()
```
Comment (Azrun) [inline code/obj/item/gun/energy.dm:2004]:
```suggestion
				UpdateIcon()
```

File: code/obj/item/gun/energy.dm [hunk 6]
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
...
+				return
+			else if (istype(W, /obj/item/light/tube) && !our_light)
+				boutput(user,SPAN_NOTICE("You place [W] inside of the barrel and redo the wiring."))
+				playsound(src, 'sound/effects/pop.ogg', 50, TRUE)
+				attach_light(W, user)
+				update_icon()
```
Comment (Azrun) [inline code/obj/item/gun/energy.dm:2010]:
```suggestion
				UpdateIcon()
```

File: code/obj/item/gun/energy.dm [hunk 7]
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
...
+			var/image/overlay_image = SafeGetOverlayImage("gun_light", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-light")
+			src.UpdateOverlays(overlay_image, "gun_light")
+		else
+			src.UpdateOverlays(null, "gun_light")
+
+		if (heat > FIRE_THRESHOLD)
+			var/image/overlay_image = SafeGetOverlayImage("gun_smoke", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-burn")
+			src.UpdateOverlays(overlay_image, "gun_smoke")
+		else if (heat > 70)
+			var/image/overlay_image = SafeGetOverlayImage("gun_smoke", 'icons/obj/items/guns/energy64x32.dmi', "makeshift-smoke")
```
Comment (Azrun) [inline code/obj/item/gun/energy.dm:2076]: Instances of `'icons/obj/items/guns/energy64x32.dmi'` can be replaced by src.icon

File: code/obj/item/gun/energy.dm [hunk 8]
```diff
@@ -1839,3 +1839,270 @@ TYPEINFO(/obj/item/gun/energy/wasp)
...
+				playsound(src, 'sound/items/Ratchet.ogg', 50, TRUE)
+				user.put_in_hand_or_drop(removed_cell)
+				return
+			else if (istype(W, /obj/item/cell) && !our_cell)
+				user.u_equip(W)
+				boutput(user,SPAN_NOTICE("You attach [W] to [src]."))
+				attach_cell(W, user)
+				return
+			else if (issnippingtool(W) && our_light)
+				boutput(user,SPAN_NOTICE("You remove [our_light] from the barrel."))
```
Comment (Azrun) [inline code/obj/item/gun/energy.dm:2000]: I'd prefer this refer to cutting away the wiring attaching the light.  Otherwise I don't understand how you remove it with a snipping implement.

**Comment by Azrun** [comment]
Should light tube adjust projectile color?

**Comment by unfunnyperson** [comment]
> > Should light tube adjust projectile color?
> 
> That would be cool but I can't figure out how to do it
Nevermind I did find a way to do this 


> > Should light tube adjust projectile color?
> 
> That would be cool but I can't figure out how to do it
figured out a way
